contrib/endpoints/src/api_manager/api_manager_impl.cc

```diff
@@ -25,43 +25,45 @@ ApiManagerImpl::ApiManagerImpl(std::unique_ptr<ApiManagerEnvInterface> env,
                                const std::string &service_config,
                                const std::string &server_config)
     : global_context_(
-          new context::GlobalContext(std::move(env), server_config)) {
-  if (!service_config.empty()) {
-    AddConfig(service_config, true);
-  }
-
+          new context::GlobalContext(std::move(env), server_config)),
+      config_manager_(nullptr),
+      service_config_(service_config),
+      config_loading_status_(
+          utils::Status(Code::UNKNOWN, "Not initialized yet")) {
   check_workflow_ = std::unique_ptr<CheckWorkflow>(new CheckWorkflow);
   check_workflow_->RegisterAll();
 }
 
-void ApiManagerImpl::AddConfig(const std::string &service_config,
-                               bool deploy_it) {
+bool ApiManagerImpl::AddConfig(const std::string &service_config,
+                               std::string *config_id) {
   std::unique_ptr<Config> config =
       Config::Create(global_context_->env(), service_config);
-  if (config != nullptr) {
-    std::string service_name = config->service().name();
-    if (global_context_->service_name().empty()) {
-      global_context_->set_service_name(service_name);
-    } else {
-      if (service_name != global_context_->service_name()) {
-        auto err_msg = std::string("Mismatched service name; existing: ") +
-                       global_context_->service_name() + ", new: " +
-                       service_name;
-        global_context_->env()->LogError(err_msg);
-        return;
-      }
-    }
-    std::string config_id = config->service().id();
-    service_context_map_[config_id] = std::make_shared<context::ServiceContext>(
-        global_context_, std::move(config));
-    // TODO: if this function is called at worker process, need to call
-    // service_context->service_control()->Init().
-    // ApiManagerImpl constructor is called at master process, not at worker
-    // process.
-    if (deploy_it) {
-      DeployConfigs({{config_id, 0}});
+  if (config == nullptr) {
+    global_context_->env()->LogError(std::string("Invalid service config: ") +
+                                     service_config);
+    return false;
+  }
+
+  std::string service_name = config->service().name();
+  if (global_context_->service_name().empty()) {
+    global_context_->set_service_name(service_name);
+  } else {
+    if (service_name != global_context_->service_name()) {
+      auto err_msg = std::string("Mismatched service name; existing: ") +
+                     global_context_->service_name() + ", new: " + service_name;
+      global_context_->env()->LogError(err_msg);
+      return false;
     }
   }
+
+  config_id->assign(config->service().id());
+
+  service_context_map_[*config_id] = std::make_shared<context::ServiceContext>(
```

Comment:
to avoid a map look up.   use

auto context_service = make_shared
context_service->Init()

then add it to the map.